In [1]:
import os
import pandas as pd
import numpy as np
import matplotlib as plt

In [2]:
df = pd.read_csv("data1.csv")
print(df.shape)
df

(4981, 11)


gender   age  hypertension  heart_disease ever_married      work_type  \
0       Male  67.0             0              1          Yes        Private   
1       Male  80.0             0              1          Yes        Private   
2     Female  49.0             0              0          Yes        Private   
3     Female  79.0             1              0          Yes  Self-employed   
4       Male  81.0             0              0          Yes        Private   
...      ...   ...           ...            ...          ...            ...   
4976    Male  41.0             0              0           No        Private   
4977    Male  40.0             0              0          Yes        Private   
4978  Female  45.0             1              0          Yes       Govt_job   
4979    Male  40.0             0              0          Yes        Private   
4980  Female  80.0             1              0          Yes        Private   

     Residence_type  avg_glucose_level   bmi   smoking_status  stroke  
0             Urban             228.69  36.6  formerly smoked       1  
1             Rural             105.92  32.5     never smoked       1  
2             Urban             171.23  34.4           smokes       1  
3             Rural             174.12  24.0     never smoked       1  
4             Urban             186.21  29.0  formerly smoked       1  
...             ...                ...   ...              ...     ...  
4976          Rural              70.15  29.8  formerly smoked       0  
4977          Urban             191.15  31.1           smokes       0  
4978          Rural              95.02  31.8           smokes       0  
4979          Rural              83.94  30.0           smokes       0  
4980          Urban              83.75  29.1     never smoked       0  

[4981 rows x 11 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4981 entries, 0 to 4980
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   gender             4981 non-null   object 
 1   age                4981 non-null   float64
 2   hypertension       4981 non-null   int64  
 3   heart_disease      4981 non-null   int64  
 4   ever_married       4981 non-null   object 
 5   work_type          4981 non-null   object 
 6   Residence_type     4981 non-null   object 
 7   avg_glucose_level  4981 non-null   float64
 8   bmi                4981 non-null   float64
 9   smoking_status     4981 non-null   object 
 10  stroke             4981 non-null   int64  
dtypes: float64(3), int64(3), object(5)
memory usage: 428.2+ KB


In [4]:
df = df.rename(columns={'Residence_type': 'residence_type'})

In [5]:
df.describe()

age  hypertension  heart_disease  avg_glucose_level  \
count  4981.000000   4981.000000    4981.000000        4981.000000   
mean     43.419859      0.096165       0.055210         105.943562   
std      22.662755      0.294848       0.228412          45.075373   
min       0.080000      0.000000       0.000000          55.120000   
25%      25.000000      0.000000       0.000000          77.230000   
50%      45.000000      0.000000       0.000000          91.850000   
75%      61.000000      0.000000       0.000000         113.860000   
max      82.000000      1.000000       1.000000         271.740000   

               bmi       stroke  
count  4981.000000  4981.000000  
mean     28.498173     0.049789  
std       6.790464     0.217531  
min      14.000000     0.000000  
25%      23.700000     0.000000  
50%      28.100000     0.000000  
75%      32.600000     0.000000  
max      48.900000     1.000000

In [6]:
print("Пол:")
print(df["gender"].unique())
print("Бывал ли в браке:")
print(df["ever_married"].unique())
print("Работа:")
print(df["work_type"].unique())
print("Тип проживания:")
print(df["residence_type"].unique())
print("Курил ли:")
print(df["smoking_status"].unique())

Пол:
['Male' 'Female']
Бывал ли в браке:
['Yes' 'No']
Работа:
['Private' 'Self-employed' 'Govt_job' 'children']
Тип проживания:
['Urban' 'Rural']
Курил ли:
['formerly smoked' 'never smoked' 'smokes' 'Unknown']


In [7]:
print(df[df["work_type"] == "Private"].shape)
print(df[df["smoking_status"] == "Unknown"].shape)

(2860, 11)
(1500, 11)


2800+ респондентов из 5000 не указали своё место работы. Кто-то из них мог вообще не работать, поэтому мы не будем учитывать данные данного столбца. 1500 людей не уточнили курили они или нет.

In [8]:
del df['work_type']

In [9]:
print("Курят:", df[df["smoking_status"] == "smokes"].shape[0])
print("Раньше курили:", df[df["smoking_status"] == "formerly smoked"].shape[0])
print("Никогда не курили:", df[df["smoking_status"] == "never smoked"].shape[0])

Курят: 776
Раньше курили: 867
Никогда не курили: 1838


Тех кто курит, мы будем кодировать числом 1, бывших курильщиков -- 0.5, а тех кто никогда не курил числом 0. Для воздержавшихся от ответа мы вставим среднее взвешенное значение: $\frac{776 \cdot 1 + 867 \cdot 0.5 + 1838 \cdot 0}{776+867+1838} \approx 0.347$

$\textbf{Нормализация продолжается:}$

In [10]:
df.loc[df["smoking_status"] == "smokes", "smoking_status"] = 1
df.loc[df["smoking_status"] == "formerly smoked", "smoking_status"] = 0.5
df.loc[df["smoking_status"] == "never smoked", "smoking_status"] = 0
df.loc[df["smoking_status"] == "Unknown", "smoking_status"] = 0.347

df.loc[df["gender"] == "Male", "gender"] = 1
df.loc[df["gender"] == "Female", "gender"] = 0

df.loc[df["ever_married"] == "Yes", "ever_married"] = 1
df.loc[df["ever_married"] == "No", "ever_married"] = 0

df.loc[df["residence_type"] == "Urban", "residence_type"] = 1 
df.loc[df["residence_type"] == "Rural", "residence_type"] = 0
df = df.rename(columns={'residence_type': 'urban_dweller'})


In [11]:
df["age"] = df["age"] / 100
df["avg_glucose_level"] = df["avg_glucose_level"] / 280
df["bmi"] = df["bmi"] / 50

In [12]:
df

gender   age  hypertension  heart_disease ever_married urban_dweller  \
0         1  0.67             0              1            1             1   
1         1  0.80             0              1            1             0   
2         0  0.49             0              0            1             1   
3         0  0.79             1              0            1             0   
4         1  0.81             0              0            1             1   
...     ...   ...           ...            ...          ...           ...   
4976      1  0.41             0              0            0             0   
4977      1  0.40             0              0            1             1   
4978      0  0.45             1              0            1             0   
4979      1  0.40             0              0            1             0   
4980      0  0.80             1              0            1             1   

      avg_glucose_level    bmi smoking_status  stroke  
0              0.816750  0.732            0.5       1  
1              0.378286  0.650              0       1  
2              0.611536  0.688              1       1  
3              0.621857  0.480              0       1  
4              0.665036  0.580            0.5       1  
...                 ...    ...            ...     ...  
4976           0.250536  0.596            0.5       0  
4977           0.682679  0.622              1       0  
4978           0.339357  0.636              1       0  
4979           0.299786  0.600              1       0  
4980           0.299107  0.582              0       0  

[4981 rows x 10 columns]